lsp-dart.el

```diff
@@ -80,6 +80,154 @@ Defaults to 0.9"
+  "When set to non-nil, dart/textDocument/publishFlutterOutline notifications will
+be sent with Flutter outline information for open files. Defaults to t"
+  :type 'boolean
+  :group 'lsp-dart
+  :package-version '(lsp-mode . "6.3"))
+
+(defcustom lsp-dart-outline-position-params
+  `((side . ,treemacs-position)
+    (slot . 2)
+    (window-width . ,treemacs-width))
+  "The outline tree position params.
+Defaults to side following treemacs default."
+  :type 'list
+  :group 'lsp-dart
+  :package-version '(lsp-mode . "6.3"))
+
+(defcustom lsp-dart-flutter-outline-position-params
+  `((side . ,treemacs-position)
+    (slot . 2)
+    (window-width . ,treemacs-width))
+  "The Flutter outline tree position params.
+Defaults to side following treemacs default."
+  :type 'list
+  :group 'lsp-dart
+  :package-version '(lsp-mode . "6.3"))
+
+(defun lsp-dart--outline-kind->icon (kind)
+  "Maps an outline KIND to a treemacs icon symbol.
+Kinds from https://github.com/dart-lang/sdk/blob/master/pkg/analysis_server/tool/spec/generated/java/types/ElementKind.java"
+  (pcase kind
+    ("CLASS" 'class)
+    ("CLASS_TYPE_ALIAS" 'class)
+    ("COMPILATION_UNIT" 'document)
+    ("FIELD" 'field)
+    ("METHOD" 'method)
+    ("CONSTRUCTOR" 'namespace)
+    ("CONSTRUCTOR_INVOCATION" 'namespace)
+    ("GETTER" 'property)
+    ("SETTER" 'property)
+    ("TOP_LEVEL_VARIABLE" 'constant)
+    ("FUNCTION" 'method)
+    ("FUNCTION_INVOCATION" 'method)
+    ("FUNCTION_TYPE_ALIAS" 'method)
+    ("LABEL" 'number)
+    ("LIBRARY" 'template)
+    ("EXTENSION" 'interface)
+    ("LOCAL_VARIABLE" 'field)
+    ("MIXIN" 'interface)
+    ("PARAMETER" 'string)
+    ("TYPE_PARAMETER" 'string)
+    ("UNIT_TEST_GROUP" 'structure)
+    ("UNIT_TEST_TEST" 'method)
+    ("ENUM" 'enumerator)
+    ("ENUM_CONSTANT" 'enumitem)
+    ("NEW-INSTANCE" 'flutter)))
+
+(defun lsp-dart--outline-tree-ret-action (uri range)
+  "Build the ret action for and item in the outline tree view.
+URI is the source of the item.
+RANGE is the range of positions to where this item should point."
+  (interactive)
+  (lsp-treemacs--open-file-in-mru (lsp--uri-to-path uri))
+  (goto-char (lsp--position-to-point (gethash "start" range)))
+  (run-hooks 'xref-after-jump-hook))
+
+(defun lsp-dart--outline->tree (uri items)
+  "Builds a outline tree.
+URI is the source of the outline.
+ITEMS is the outline items data."
+  (seq-map (-lambda ((&hash "children"
+                            "element" (&hash "kind" "name" "parameters" "range")))
+             (let ((label (concat name (when parameters
+                                         (propertize (concat " " parameters)
+                                                     'face 'lsp-lens-face)))))
+               (list :key label
+                     :label label
+                     :icon (lsp-dart--outline-kind->icon kind)
+                     :children (unless (seq-empty-p children)
+                                 (lsp-dart--outline->tree uri children))
```

Comment:
in general, you can pass a function to :children to make the conversion lazy. You may skip this one if you do not expect this function to be called often with big data.